lib/errors/index.ts

```diff
@@ -0,0 +1,94 @@
+import * as http from 'http';
+import * as rawErrors from './arsenalErrors';
+import * as types from './types';
+
+export * from './types';
+
+/** Mapping used to determine an error type. */
+export type Is = { [Name in types.Name]: boolean };
+/** Mapping of all possible Errors */
+export type Errors = { [Property in keyof types.Names]: ArsenalError };
+
+const entries = Object.entries(rawErrors);
+
+// This contains some metaprog. Be careful.
+// Proxy can be found on MDN.
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy
+const createIs = (type: types.Name) => {
+    const get = (_: {}, value: string | symbol) => type === value;
+    return new Proxy({}, { get }) as Is;
+};
+
+export class ArsenalError extends Error {
+    /** HTTP status code. Example: 401, 403, 500, ... */
+    #code: number;
+    /** Text description of the error. */
+    #description: string;
+    /** Type of the error. Belongs to errors.types. */
+    #type: types.Name;
+    /** Object used to determine the error type.
+     * Example: error.is.InternalError */
+    #is: Is;
+
+    private constructor(type: types.Name, code: number, description: string) {
+        super(type);
+        this.#code = code;
+        this.#description = description;
+        this.#type = type;
+        this.#is = createIs(type);
+    }
+
+    /** Output the error as a JSON string */
+    toString() {
+        const errorType = this.message;
+        const errorMessage = this.#description;
+        return JSON.stringify({ errorType, errorMessage });
+    }
+
+    /** Write the error in an HTTP response */
+    writeResponse(res: http.ServerResponse) {
+        res.writeHead(this.#code);
+        const asStr = this.toString();
+        res.end(asStr);
+    }
+
+    /** Clone the error with a new description.*/
+    customizeDescription(description: string): ArsenalError {
+        const type = this.#type;
+        const code = this.#code;
+        return new ArsenalError(type, code, description);
+    }
+
+    get is() {
+        return this.#is;
+    }
+
+    get code() {
+        return this.#code;
+    }
+
+    get description() {
+        return this.#description;
+    }
+
+    get type() {
+        return this.#type;
+    }
+
+    /** Generate all possible errors. An instance is created by default. */
+    static errors() {
```

Comment:
I would move this out of the class personally, Could also have a return type or does TS auto find it?

Reply:
The type is inferred from the return statement - the engine sees a map of `types.Name` to `ArsenalError` in the reduce, it just yells at us in the base case because `{}` can be almost anything, so we cast it to `Errors`.

It needs to be in the class because that's the only place we want to allow calling the `ArsenalError` constructor - all other errors are either specific errors, or created through the `customizeDescription` call, which is in the class.

Reply:
oh right, the private constructor, that's fine then.